aptos-move/block-executor/src/code_cache_global_manager.rs

```diff
@@ -0,0 +1,599 @@
+    environment: ExplicitSyncWrapper<Option<AptosEnvironment>>,
+}
+
+impl<T, K, DC, VC, E> ModuleCacheManager<T, K, DC, VC, E>
+where
+    T: Debug + Eq,
+    K: Hash + Eq + Clone,
+    VC: Deref<Target = Arc<DC>>,
+    E: WithSize,
+{
+    /// Returns a new instance of [ModuleCacheManager] in a [State::Done] state with uninitialized
+    /// current value.
+    #[allow(clippy::new_without_default)]
+    pub fn new() -> Self {
+        Self {
+            state: Mutex::new(State::Done(None)),
+            module_cache: Arc::new(RwLock::new(GlobalModuleCache::empty())),
+            environment: ExplicitSyncWrapper::new(None),
+        }
+    }
+
+    /// If state is [State::Done], sets the state to [State::Ready] with the current value and
+    /// returns true. Otherwise, raises an alert and returns false. Additionally, synchronizes
+    /// module and environment caches based on the provided previous value.
+    pub fn mark_ready(&self, previous: Option<&T>, current: Option<T>) -> bool {
+        let mut state = self.state.lock();
+
+        if let State::Done(recorded_previous) = state.deref() {
+            // If the state is done, but the values do not exist or do not match, we flush global
+            // caches because they execute on top of unknown state (or on top of some different to
+            // the previous state).
+            if !recorded_previous
+                .as_ref()
+                .is_some_and(|r| previous.is_some_and(|p| r == p))
+            {
+                if let Some(environment) = self.environment.acquire().as_ref() {
+                    environment
+                        .runtime_environment()
+                        .flush_struct_name_and_info_caches();
+                }
+                self.module_cache.write().flush_unsync();
+            }
+
+            *state = State::Ready(current);
+            true
+        } else {
+            // We are not in the done state, this is an error.
+            alert_or_println!(
+                "Unable to mark ready, state: {:?}, previous: {:?}, current: {:?}",
+                state,
+                previous,
+                current
+            );
+            false
+        }
+    }
+
+    /// When in [State::Ready], runs different checks on cached modules and environment:
+    ///   1. If the environment is not initialized, or is different from the one in storage, it is
+    ///      re-initialized, and module caches are flushed.
+    ///   2. If too many struct names have been cached in re-indexing map in runtime environment,
+    ///      struct type caches and module caches are flushed.
+    ///   3. If module cache size is too large (in bytes), it is flushed.
+    /// The final environment and module caches are returned.
+    pub fn check_ready_and_get_caches(
+        &self,
+        storage_environment: AptosEnvironment,
+        config: &BlockExecutorModuleCacheLocalConfig,
+    ) -> Result<
```

Comment:
Can't we have a race here?
some block executor calls 
`check_ready_and_get_caches`
returns an Arc to cache, and environment, at at that time state was correct.

Now, before the corresponding block execution (in wrapper mod) gets to mark_executing, a different block execution goes through a full cycle on the manager, bringing it back to ready but with a different hash (either flush, or maybe it had a same parent, so executed and applied the updates). Not sure if the environment is shared, but code cache is shared via the outside Arc that is cloned here, so by the time we lock the RWLock and provide the guard, the cache has changed from underneath us..

Would be good to have a unit test for this scenario as well, and I see two ways to fix:
(a)  Maybe check_ready (and get caches) should be like CAS and already change status to executing, or mark_executing can double check the hash, 
(b) we can already provide lock guard here, but then I think the manager can be redesigned a little - i.e. the hash (and maybe even the environment) can all live inside the RWLock. check_ready would be equivalent to getting a read lock, checking hash matches, and if it matches returning the guard (and environment). If it doesn't match - try to acquire write lock, if successfully acquired flush the whole thing and retry (empty matches anything), if not able to acquire someone else is running and just run the local one with empty cache (or wait). 

@ziaptos what do you think? I think you would prefer the redesign since now we used the lock with guard anyway? I do think it's nice, since manager now also becomes cleaner instead of members having 3 different synchronization mechanisms (mutex, rwlock and explicitsyncwrapper).

Reply:
I think what we really need is to pass down the manager to block executor so we can keep caches in sync, and lock the manager per block execution?

Reply:
yes, just responded to the other comment - that seems the simplest way to do it cleanly from the current state. If we pass it down and do the checks in block executor, the whole manager synchronization logic (env, hash, flushing) can be under one RWLock.

Reply:
Actually, we need two things: 
1. Global lock on execution (per block)
2. No interior mutability of global module cache, pass & everywhere.
3. Modifications to cache are through lock on execution

Reply:
Yes!

Reply:
hm, if it is read locked during block execution, and after acquiring read lock we check the invariants - it is guaranteed nobody can modify from under us, right?
modifications would happen via write lock at the end of block execution - it can simply upgrade the read lock and re-check invariants.
flushing can happen if invariants did not match and write lock was acquired.

Reply:
I meant that it can be exclusively locked during block execution? And we just set the state as `read`/`write` for the sections we care about? I am not sure about upgrades, cannot it cause a deadlock?

Reply:
So what we can do:
1. Have a Mutex<Manager>, lock before execution
2. Manager checks invariants, in a write mode, prefetches, etc.
3. Set to read-only, execute txns
4. Set to write to add modules from mvhashmap, etc.
5. When done, unlock.

Reply:
@georgemitenkov can you write here what we decided / you ended up implementing, for documenting if we look at the PR later.

Reply:
Implemented, see last commit. We have a Mutex<manager> which we try lock (doing all checks) to get a guard. Guard is passed to block executor and from that we get the module cache and environment.